Initial support for OCI key manager

pjcolp opened the pull request:
This patchset introduces OCIKeyManager, a new key manager backed by OCI Key Vault.

Every effort was made to match the existing code structure, but I'm open to suggestions on how certain bits of code may be separated out (for example, the KeyInfo and KeyAttribute structs).

pjcolp committed:
oci: Add OCI key manager to the list of available key managers

This enables the use of the OCI key manager.

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
ociclient: Initialize OCI client

OCI uses different clients for different functionality. To perform operations
on a key, the vaults client is used. To actually get the contents of a key,
the secrets client is used. Create an instance of each client.

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
keys: Add OCI KeyAttributes members

The KeyAttributes struct is a union of all needed fields (for example, KMIP
fields are present). Add OCI specific members to the KeyAttributes struct as
well.

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
keys: Add OCI KeyInfo members

The KeyInfo struct is a union of all needed fields (for example, KMIP fields
are present). Add OCI specific members to the KeyInfo struct as well.

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
oci_key_manager: Add TransferKey() functionality

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
oci_key_manager: Add CreateKey() functionality

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
ociclient: Add DeleteKey() functionality

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
oci_key_manager: Add DeleteKey() functionality

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
oci: Add initial OCI client and key manager

Note that (for now) these only implement stub functionality. Full
functionality will be added in later commits.

Update go.mod and go.sum with new dependencies for OCI SDK.

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

arvind5 commented on keymanager/oci_key_manager.go:
```diff
@@ -0,0 +1,103 @@
+func (om *OCIManager) DeleteKey(keyAttributes *model.KeyAttributes) error {
+	err := om.client.DeleteKey(keyAttributes.OciSecretId)
+	if err != nil {
+		log.Errorf("Error while deleting key: %s", err.Error())
```
use errors.wrap() for consistency.

mythi replied:
I'd suggest to re-consider the usage of `github.com/pkg/errors` since it's archived and not maintained. In general, projects are trying to move away from it and use error wrapping from the std library.

pjcolp replied:
I'll update it to errors.Wrap() for now. I was just copying the code from   vault_key_manager.go:DeleteKey(). If we want to move the std library implementation, I would suggest doing that as a separate patch(set) that updates error usage everywhere.

I'm also not 100% sure what wrapping from the std library you want to do -- I don't see an errors.Wrap() function there. Do you mean using fmt.Errorf() or errors.Join()?

mythi replied:
I was thinking`%w`  with `err`

arvind5 commented on ociclient/ociclient.go:
```diff
@@ -0,0 +1,153 @@
+	req := vault.CreateSecretRequest{
+		CreateSecretDetails: vault.CreateSecretDetails{
+			// Required
+			CompartmentId: common.String(keyAttributes.OciCompartmentId),
```
just a thought, would it be better to pass only the required arguments to this function instead of keyAttributes struct, just to keep the client independent of keyAttributes implementation.

pjcolp replied:
I went back and forth on that one. In the end, I again decided to model it after the Vault implementation. vaultclient.go:CreateKey() also takes a KeyAttributes struct.

I'm happy to change it to take the required arguments, though. I can also (in a separate pull request) also similarly update the vaultclient code.

arvind5 replied:
checked vaultclient implementation, its bit different in the sense that it sends entire KeyAttributes struct to backend. I would let @hmgowda comment on updating vaultclient code.

arvind5 commented on ociclient/ociclient.go:
```diff
@@ -0,0 +1,153 @@
+		return nil, err
+	}
+
+	return []byte(decodedStr), nil
```
DecodeString returns []byte so no need to typecast.

arvind5 commented on ociclient/ociclient.go:
```diff
@@ -0,0 +1,153 @@
+	// Send the request using the secrets client.
+	resp, err := oc.sc.GetSecretBundle(context.Background(), req)
+	if err != nil {
+		return nil, err
```
good to wrap the error with user-friendly message.

arvind5 commented on service/key.go:
```diff
@@ -53,7 +53,8 @@ func (svc service) CreateKey(_ context.Context, keyCreateReq model.KeyRequest) (
 	var createdKey *model.KeyResponse
-	if keyCreateReq.KeyInfo.KeyData == "" && keyCreateReq.KeyInfo.KmipKeyID == "" {
+	if keyCreateReq.KeyInfo.KeyData == "" &&
+		(keyCreateReq.KeyInfo.KmipKeyID == "" || keyCreateReq.KeyInfo.OciSecretId == "") {
```
assuming OciSecretId is meant for Register operation, OR should be replaced with AND.

pjcolp replied:
Ah yes, good catch.

arvind5 commented on model/key.go:
```diff
@@ -60,6 +60,21 @@ type KeyInfo struct {
 	KmipKeyID string `json:"kmip_key_id,omitempty"`
+	// The OCID of the compartment where you want to create the secret.
+	// example: ocid1.test.oc1..<unique_ID>EXAMPLE-compartmentId-Value
+	OciCompartmentId string `json:"oci_compartment_id,omitempty"`
```
do we need to return all these fields in response? if not, then we can keep these in separate struct.

pjcolp replied:
Yeah, that's what I was thinking too. I just did this for now since the Vault and KMIP stuff was also all mixed in there. If we want to separate them out, then I can do a patch that first pulls out the Vault and KMIP stuff into separate structs, then change this patch to also be in a separate struct.

arvind5 replied:
we only have KmipKeyID field for KMIP which is needed in both request and response (none specific to vault), hence kept KmipKeyID in KeyInfo struct which is shared between request and response. If the OCI fields also need to be kept in both request and response, then we can leave them in KeyInfo struct.

pjcolp replied:
While separating out the OCI stuff, I was going over the other code. Both `KeyData` and `KmipKeyID` are actually both only used with KeyRequest as well. They are also in the KeyAttributes struct. Is the intent is that KeyAttributes and KeyInfo should roughly match?

If so, one option for the OCI code is to just put `OciSecretId` in KeyInfo and KeyAttributes, as that's the only one needed in both. Then I can put the rest of the OCI fields into their own struct and just include that in the KeyRequest struct

arvind5 replied:
updates looks good.

pjcolp committed:
oci_key_manager: Add RegisterKey() functionality

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
ociclient: Add GetKey() functionality

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp commented:
I've pushed an updated version that I believe addresses all the comments here. Please have a look and let me know if there are any other changes required.

arvind5 commented on ociclient/ociclient.go:
```diff
@@ -0,0 +1,152 @@
+	return nil
+}
+
+func (oc *ociClient) CreateKey(OciCompartmentId, OciKeyId, OciSecretName, OciVaultId string, OciSecretId *string) error {
```
update function signature to return secretId instead of passing output variable for storing secretId.

pjcolp replied:
Ah, yeah. That makes sense, will do.

arvind5 commented on ociclient/ociclient.go:
```diff
@@ -0,0 +1,152 @@
+		return errors.Wrapf(err, "Failed to delete key '%s' from oci server", secretId)
+	}
+
+	// TODO: Check response for error.
```
any plans to address this in this PR?

arvind5 commented on ociclient/ociclient.go:
```diff
@@ -0,0 +1,152 @@
+	re := regexp.MustCompile(`\{ Content=([-A-Za-z0-9+/]*) \}`)
+	strs := re.FindStringSubmatch(fmt.Sprintf("%s", resp.SecretBundle.SecretBundleContent))
+	if strs == nil || len(strs) != 2 {
+		return nil, errors.New("could not match regular expression")
```
can we improve error message? `could not parse secret bundle` or `could not extract secret`?

pjcolp replied:
Will do.

pjcolp committed:
ociclient: Add CreateKey() functionality

Currently just support the "BYTES" secret type (and use "BYTES_512").
The "SSH_KEY" and "PASSPHRASE" types will be added later, as will support
for "BYTES_1024".

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp commented:
I've pushed another update with the latest set of fixes.

arvind5 commented:
> I've pushed another update with the latest set of fixes.

@pjcolp, unit tests are failing, can you please check?

pjcolp commented:
@arvind5 OK, should be working now

arvind5 commented:
> @arvind5 OK, should be working now

@pjcolp, unit tests are fixed, however the test coverage is dropped to 78%, can you please check?

pjcolp committed:
oci_key_manager: Add unit tests

Add unit tests for the OCI key manager service.

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp commented:
@arvind5 Test coverage should be back up to (almost) where it was (80.6% before to 80.5% now)

arvind5 commented:
> @arvind5 Test coverage should be back up to (almost) where it was (80.6% before to 80.5% now)

@pjcolp, thanks for improving the test coverage, however the CI is failing due to unit tests failure, can you please check?

pjcolp commented:
@arvind5 OK, should be fixed now

arvind5 commented:
> @arvind5 OK, should be fixed now

@pjcolp, thanks for fixing the unit tests, looks good to me.

hmgowda commented on config/default.go:
```diff
@@ -59,7 +59,7 @@ func DefaultConfig() *Configuration {
 			ClientToken: viper.GetString(VaultClientToken),
 		}
-	} else {
+	} else if strings.ToLower(cfg.KeyManager) == constant.KmipKeyManager {
```
Since, OCI uses the same config as Vault keymanager. It would be better if you could add  something like this to the default config.  "strings.ToLower(cfg.KeyManager) == constant.VaultKeyManager || strings.ToLower(cfg.KeyManager) == constant.OCIKeyManager" and keep the else block as is for KMIP keymanager

pjcolp replied:
OK, I've updated it. Doing this gives back the test coverage % for config as well.

hmgowda replied:
sounds good.

hmgowda commented on config/default.go:
```diff
@@ -72,5 +72,6 @@ func DefaultConfig() *Configuration {
 			RootCertificateFilePath:   viper.GetString(KmipRootCertPath),
 		}
 	}
+	// Currently we do nothing special for OCI config.
```
remove this comment once the if-else block is implemented with a check for OCI keymanager

hmgowda commented on keymanager/oci_key_manager.go:
```diff
@@ -0,0 +1,110 @@
+func (om *OCIManager) CreateKey(keyRequest *model.KeyRequest) (*model.KeyAttributes, error) {
+	if keyRequest.OciInfo.CompartmentId == "" || keyRequest.OciInfo.KeyId == "" ||
+		keyRequest.OciInfo.SecretName == "" || keyRequest.OciInfo.VaultId == "" {
+		return nil, errors.New("Missing oci_compartment_id, oci_key_id, oci_secret_name, or oci_vault_id")
```
If the request to create a key is being made, how will the user get all of this information prior to the key being created.

pjcolp replied:
The user first needs to setup an Oracle Cloud account and create a master encryption key and a vault (which goes into a compartment). Once that's done, they can get the OCIDs from them. The secret name is something chosen by the user at creation time.

I've added information about this to the README file.

pjcolp committed:
OCI: Add documentation to docs/keys.go

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

pjcolp committed:
README: Add instructions for setting up OCI

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>

hmgowda commented:
@pjcolp  we will start running regression tests on this branch soon. Can you please rebase the branch with main and resolve all conflicts?

hmgowda commented on README.md:
```diff
@@ -192,6 +212,11 @@ On Linux, follow the steps below to install the KBS:
+
+   There is no special configuration required to use OCI Vault.
+
 3. Optionally, configure a proxy setting.
```
In step 4, can you add a note saying if OCI is the keymanager, then the config volume needs to mapped as well.  And, also provide a same docker run command for OCI KMS

docker run -d --restart unless-stopped --name kbs --env-file kbs.env -p 5443:9443 -v /etc/kbs/certs:/etc/kbs/certs -v /opt/kbs:/opt/kbs -v /home/<user>/.oci:/home/kbs/.oci trustauthority/key-broker-service:v1.2.0

pjcolp replied:
OK, I've added it in

hmgowda commented on service/key.go:
```diff
@@ -53,7 +54,9 @@ func (svc service) CreateKey(_ context.Context, keyCreateReq model.KeyRequest) (
+	if keyCreateReq.KeyInfo.KeyData == "" &&
+		keyCreateReq.KeyInfo.KmipKeyID == "" &&
+		(keyCreateReq.OciInfo == nil || keyCreateReq.OciInfo.SecretId == "") {
 
```
@pjcolp can you please add validation checks to the all the fields that are required in OciInfo block?
Here we are just add nil checks, we would like to make sure that the inputs are sanitized as well before using them in service level logic.

pjcolp replied:
@hmgowda I've added a regex check on the input strings to make sure they only contain valid characters and are in the correct format for an OCID.

Note: Neither the Vault client nor the KMIP client sanitize their input strings, but that's the subject of a different patchset.

pjcolp committed:
oci: Sanitize input strings

Check that the input strings used for OCI contain a valid structure and
only contain valid characters.

Signed-off-by: Patrick Colp <patrick.colp@oracle.com>
